rct_optimizations/include/rct_optimizations/dh_chain_kinematic_calibration.h

```diff
@@ -0,0 +1,165 @@
+
+namespace rct_optimizations
+{
+struct KinematicCalibrationProblem2D3D
+{
+  KinematicCalibrationProblem2D3D(DHChain camera_chain_, DHChain target_chain_)
+    : camera_chain(std::move(camera_chain_))
+    , target_chain(std::move(target_chain_))
+    , camera_mount_to_camera_guess(Eigen::Isometry3d::Identity())
+    , target_mount_to_target_guess(Eigen::Isometry3d::Identity())
+    , camera_base_to_target_base_guess(Eigen::Isometry3d::Identity())
+  {
+  }
+
+  DHChain camera_chain;
+  DHChain target_chain;
+  KinObservation2D3D::Set observations;
+  CameraIntrinsics intr;
+  Eigen::Isometry3d camera_mount_to_camera_guess;
+  Eigen::Isometry3d target_mount_to_target_guess;
+  Eigen::Isometry3d camera_base_to_target_base_guess;
+};
+
+struct KinematicCalibrationResult
+{
+  bool converged;
+  double initial_cost_per_obs;
+  double final_cost_per_obs;
+  Eigen::Isometry3d camera_mount_to_camera;
+  Eigen::Isometry3d target_mount_to_target;
+  Eigen::Isometry3d camera_base_to_target_base;
+  Eigen::MatrixX4d camera_chain_dh_offsets;
+  Eigen::MatrixX4d target_chain_dh_offsets;
+
+  // TODO: Add covariance matrix/matrices
+//  Eigen::MatrixXd covariance;
+};
+
+class DualDHChainCost2D3D
+{
+  public:
+  DualDHChainCost2D3D(const Eigen::Vector2d &obs,
+       const Eigen::Vector3d &point_in_target,
+       const CameraIntrinsics &intr,
+       const DHChain &camera_chain,
+       const DHChain &target_chain,
+       const Eigen::VectorXd &camera_chain_joints,
+       const Eigen::VectorXd &target_chain_joints)
+    : obs_(obs)
+    , target_pt_(point_in_target)
+    , intr_(intr)
+    , camera_chain_(camera_chain)
+    , target_chain_(target_chain)
+    , camera_chain_joints_(camera_chain_joints)
+    , target_chain_joints_(target_chain_joints)
+  {
+  }
+
+  template<typename T>
+  Isometry3<T> createTransform(T const *const *params, const std::size_t idx) const
+  {
+    Eigen::Map<const Vector3<T>> t(params[idx]);
+    Eigen::Map<const Vector3<T>> aa(params[idx + 1]);
+
+    Isometry3<T> result = Isometry3<T>::Identity() * Eigen::Translation<T, 3>(t);
+
+    T aa_norm = aa.norm();
+    if (aa_norm > std::numeric_limits<T>::epsilon())
+    {
+      result *= Eigen::AngleAxis<T>(aa_norm, aa.normalized());
+    }
+
+    return result;
+  }
+
+  static std::vector<double *> constructParameters(Eigen::MatrixX4d &camera_chain_dh_offsets,
+                                                   Eigen::MatrixX4d &target_chain_dh_offsets,
+                                                   Eigen::Vector3d &t_cm_to_c,
+                                                   Eigen::Vector3d &aa_cm_to_c,
+                                                   Eigen::Vector3d &t_tm_to_t,
+                                                   Eigen::Vector3d &aa_tm_to_t,
+                                                   Eigen::Vector3d &t_ccb_to_tcb,
+                                                   Eigen::Vector3d &aa_ccb_to_tcb)
```

Comment:
It would be helpful to either give the arguments more descriptive names or add some `@param` documentation listing what each one corresponds to.

e:  I see that they're documented in more detail in the `optimize` function. I still think it's a good idea to define them here too.